libs/vkd3d/shared_metadata.c

```diff
@@ -0,0 +1,51 @@
+#define VKD3D_DBG_CHANNEL VKD3D_DBG_CHANNEL_API
+#include "vkd3d_private.h"
+
+#include "winioctl.h"
+
+#define IOCTL_SHARED_GPU_RESOURCE_SET_METADATA           CTL_CODE(FILE_DEVICE_VIDEO, 4, METHOD_BUFFERED, FILE_WRITE_ACCESS)
+
+bool vkd3d_set_shared_metadata(HANDLE handle, void *buf, uint32_t buf_size)
+{
+    DWORD ret_size;
+
+    return DeviceIoControl(handle, IOCTL_SHARED_GPU_RESOURCE_SET_METADATA, buf, buf_size, NULL, 0, &ret_size, NULL);
+}
+
+#define IOCTL_SHARED_GPU_RESOURCE_GET_METADATA           CTL_CODE(FILE_DEVICE_VIDEO, 5, METHOD_BUFFERED, FILE_READ_ACCESS)
+
+bool vkd3d_get_shared_metadata(HANDLE handle, void *buf, uint32_t buf_size, uint32_t *metadata_size)
+{
+    DWORD ret_size;
+
+    bool ret = DeviceIoControl(handle, IOCTL_SHARED_GPU_RESOURCE_GET_METADATA, NULL, 0, buf, buf_size, &ret_size, NULL);
+
+    if (metadata_size)
+        *metadata_size = ret_size;
+
+    return ret;
+}
+
+#define IOCTL_SHARED_GPU_RESOURCE_OPEN                   CTL_CODE(FILE_DEVICE_VIDEO, 1, METHOD_BUFFERED, FILE_WRITE_ACCESS)
+
+HANDLE vkd3d_open_kmt_handle(HANDLE kmt_handle)
+{
+    struct
+    {
+        unsigned int kmt_handle;
+        WCHAR name[1];
```

Comment:
This `WCHAR name[1]` needs some explaining, I don't get it. Probably wrong(?) Did you mean to do like a VLA or something??

Reply:
The same IOCTL is also used for opening named shared resources, when kmt_handle is 0.  The pattern of a 1 sized array is a pattern used in the wineserver interface indicating that the allocator can instead pass in a dynamically allocated structure with more string bytes.

Reply:
Should put that in a comment.